File: android/src/main/java/com/cloudwebrtc/webrtc/SurfaceTextureRenderer.java
package com.cloudwebrtc.webrtc;

import android.graphics.SurfaceTexture;
import android.view.Surface;

import org.webrtc.EglBase;
import org.webrtc.EglRenderer;
import org.webrtc.GlRectDrawer;
import org.webrtc.RendererCommon;
import org.webrtc.ThreadUtils;
import org.webrtc.VideoFrame;

import java.util.concurrent.CountDownLatch;

import io.flutter.view.TextureRegistry;

/**
 * Display the video stream on a Surface.
 * renderFrame() is asynchronous to avoid blocking the calling thread.
 * This class is thread safe and handles access from potentially three different threads:
 * Interaction from the main app in init, release and setMirror.
 * Interaction from C++ rtc::VideoSinkInterface in renderFrame.
 * Interaction from SurfaceHolder lifecycle in surfaceCreated, surfaceChanged, and surfaceDestroyed.
 */
public class SurfaceTextureRenderer extends EglRenderer {
  // Callback for reporting renderer events. Read-only after initilization so no lock required.
  private RendererCommon.RendererEvents rendererEvents;
  private final Object layoutLock = new Object();
  private boolean isRenderingPaused;
  private boolean isFirstFrameRendered;
  private int rotatedFrameWidth;
  private int rotatedFrameHeight;
  private int frameRotation;

  /**
   * In order to render something, you must first call init().
   */
  public SurfaceTextureRenderer(String name) {
    super(name);
  }

  public void init(final EglBase.Context sharedContext,
                   RendererCommon.RendererEvents rendererEvents) {
    init(sharedContext, rendererEvents, EglBase.CONFIG_PLAIN, new GlRectDrawer());
  }

  /**
   * Initialize this class, sharing resources with |sharedContext|. The custom |drawer| will be used
   * for drawing frames on the EGLSurface. This class is responsible for calling release() on
   * |drawer|. It is allowed to call init() to reinitialize the renderer after a previous
   * init()/release() cycle.
   */
  public void init(final EglBase.Context sharedContext,
                   RendererCommon.RendererEvents rendererEvents, final int[] configAttributes,
                   RendererCommon.GlDrawer drawer) {
    ThreadUtils.checkIsOnMainThread();
    this.rendererEvents = rendererEvents;
    synchronized (layoutLock) {
      isFirstFrameRendered = false;
      rotatedFrameWidth = 0;
      rotatedFrameHeight = 0;
      frameRotation = -1;
    }
    super.init(sharedContext, configAttributes, drawer);
  }
  @Override
  public void init(final EglBase.Context sharedContext, final int[] configAttributes,
                   RendererCommon.GlDrawer drawer) {
    init(sharedContext, null /* rendererEvents */, configAttributes, drawer);
  }
  /**
   * Limit render framerate.
   *
   * @param fps Limit render framerate to this value, or use Float.POSITIVE_INFINITY to disable fps
   *            reduction.
   */
  @Override
  public void setFpsReduction(float fps) {
    synchronized (layoutLock) {
      isRenderingPaused = fps == 0f;
    }
    super.setFpsReduction(fps);
  }
  @Override
  public void disableFpsReduction() {
    synchronized (layoutLock) {
      isRenderingPaused = false;
    }
    super.disableFpsReduction();
  }
  @Override
  public void pauseVideo() {
    synchronized (layoutLock) {
      isRenderingPaused = true;
    }
    super.pauseVideo();
  }
  // VideoSink interface.
  @Override
  public void onFrame(VideoFrame frame) {
    if(surface == null) {
      producer.setSize(frame.getRotatedWidth(),frame.getRotatedHeight());
      surface = producer.getSurface();
      createEglSurface(surface);
    }
    updateFrameDimensionsAndReportEvents(frame);
    super.onFrame(frame);
  }

  private Surface surface = null;

  private TextureRegistry.SurfaceProducer producer;

  public void surfaceCreated(final TextureRegistry.SurfaceProducer producer) {
    ThreadUtils.checkIsOnMainThread();
    this.producer = producer;
    this.producer.setCallback(
            new TextureRegistry.SurfaceProducer.Callback() {
              @Override
              public void onSurfaceAvailable() {
                // Do surface initialization here, and draw the current frame.
              }

              @Override
              public void onSurfaceCleanup() {
                surfaceDestroyed();
              }
            }
    );
  }

  public void surfaceDestroyed() {
    ThreadUtils.checkIsOnMainThread();
    final CountDownLatch completionLatch = new CountDownLatch(1);
    releaseEglSurface(completionLatch::countDown);
    ThreadUtils.awaitUninterruptibly(completionLatch);
    surface = null;
  }

  // Update frame dimensions and report any changes to |rendererEvents|.
  private void updateFrameDimensionsAndReportEvents(VideoFrame frame) {
    synchronized (layoutLock) {
      if (isRenderingPaused) {
        return;
      }
      if (!isFirstFrameRendered) {
        isFirstFrameRendered = true;
        if (rendererEvents != null) {
          rendererEvents.onFirstFrameRendered();
        }
      }
      if (rotatedFrameWidth != frame.getRotatedWidth()
              || rotatedFrameHeight != frame.getRotatedHeight()
              || frameRotation != frame.getRotation()) {
        if (rendererEvents != null) {
          rendererEvents.onFrameResolutionChanged(
                  frame.getBuffer().getWidth(), frame.getBuffer().getHeight(), frame.getRotation());
        }
        rotatedFrameWidth = frame.getRotatedWidth();
        rotatedFrameHeight = frame.getRotatedHeight();
        producer.setSize(rotatedFrameWidth, rotatedFrameHeight);
        frameRotation = frame.getRotation();
      }
    }
  }
}

File: android/src/main/kotlin/com/instrumentisto/medea_flutter_webrtc/SurfaceTextureRenderer.kt
package com.instrumentisto.medea_flutter_webrtc

import android.graphics.SurfaceTexture
import java.util.concurrent.CountDownLatch
import org.webrtc.*
import org.webrtc.RendererCommon.GlDrawer
import org.webrtc.RendererCommon.RendererEvents

/** Displays the video stream on a `Surface`. */
class SurfaceTextureRenderer(name: String) : EglRenderer(name) {
  // Callback for reporting renderer events. Read-only after initialization,
  // so no lock is required.
  private var rendererEvents: RendererEvents? = null
  private val layoutLock = Any()

  @Volatile private var isRenderingPaused = false
  private var isFirstFrameRendered = false
  private var rotatedFrameWidth = 0
  private var rotatedFrameHeight = 0
  private var frameRotation = 0
  private var texture: SurfaceTexture? = null

  /**
   * Initialize this class, sharing resources with |sharedContext|. The custom |drawer| will be used
   * for drawing frames on the `EGLSurface`. This class is responsible for calling `release()` on
   * the |drawer|. It's allowed to call `init()` to reinitialize the renderer after the previous
   * `init()`/`release()` cycle.
   */
  @JvmOverloads
  fun init(
      sharedContext: EglBase.Context?,
      rendererEvents: RendererEvents?,
      configAttributes: IntArray? = EglBase.CONFIG_PLAIN,
      drawer: GlDrawer? = GlRectDrawer()
  ) {
    ThreadUtils.checkIsOnMainThread()
    this.rendererEvents = rendererEvents
    synchronized(layoutLock) {
      isFirstFrameRendered = false
      rotatedFrameWidth = 0
      rotatedFrameHeight = 0
      frameRotation = -1
    }
    super.init(sharedContext, configAttributes, drawer)
  }

  override fun init(sharedContext: EglBase.Context?, configAttributes: IntArray, drawer: GlDrawer) {
    init(sharedContext, null, configAttributes, drawer)
  }

  /**
   * Limit render framerate.
   *
   * @param fps Limit render framerate to this value, or use [Float.POSITIVE_INFINITY] to disable
   * FPS reduction.
   */
  override fun setFpsReduction(fps: Float) {
    isRenderingPaused = fps == 0f
    super.setFpsReduction(fps)
  }

  override fun disableFpsReduction() {
    isRenderingPaused = false
    super.disableFpsReduction()
  }

  override fun pauseVideo() {
    isRenderingPaused = true
    super.pauseVideo()
  }

  override fun onFrame(frame: VideoFrame) {
    synchronized(layoutLock) {
      if (isRenderingPaused) {
        return
      }
      if (!isFirstFrameRendered) {
        isFirstFrameRendered = true
        if (rendererEvents != null) {
          rendererEvents!!.onFirstFrameRendered()
        }
      }
      if (rotatedFrameWidth != frame.rotatedWidth ||
          rotatedFrameHeight != frame.rotatedHeight ||
          frameRotation != frame.rotation) {
        if (rendererEvents != null) {
          rendererEvents!!.onFrameResolutionChanged(
              frame.buffer.width, frame.buffer.height, frame.rotation)
        }
        rotatedFrameWidth = frame.rotatedWidth
        rotatedFrameHeight = frame.rotatedHeight
        texture!!.setDefaultBufferSize(rotatedFrameWidth, rotatedFrameHeight)
        frameRotation = frame.rotation
      }
    }
    super.onFrame(frame)
  }

  fun surfaceCreated(texture: SurfaceTexture) {
    ThreadUtils.checkIsOnMainThread()
    this.texture = texture
    createEglSurface(texture)
  }

  fun surfaceDestroyed() {
    ThreadUtils.checkIsOnMainThread()
    val completionLatch = CountDownLatch(1)
    releaseEglSurface { completionLatch.countDown() }
    ThreadUtils.awaitUninterruptibly(completionLatch)
  }
}
